File: JavaApp/src/main/java/com/charlezz/javaapp/feature/local/PersonAdapter.java
package com.charlezz.javaapp.feature.local;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.charlezz.javaapp.BR;
import com.charlezz.javaapp.R;
import com.charlezz.javaapp.databinding.ViewPersonBinding;
import com.charlezz.javaapp.feature.base.BaseAdapter;
import com.charlezz.javaapp.feature.base.BaseViewHolder;

public class PersonAdapter extends BaseAdapter<Person,PersonAdapter.PersonViewHolder> {
    private static DiffUtil.ItemCallback<Person> diffCallback = new DiffUtil.ItemCallback<Person>() {
        @Override
        public boolean areItemsTheSame(@NonNull Person oldItem, @NonNull Person newItem) {
            return  oldItem.getId() == newItem.getId() &&
                    oldItem.getName().equals(newItem.getBirth()) &&
                    oldItem.getName().equals(newItem.getName());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Person oldItem, @NonNull Person newItem) {
            return oldItem.equals(newItem);
        }
    };

    protected PersonAdapter() {
        super(diffCallback);
    }

    @NonNull
    @Override
    public PersonViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new PersonViewHolder((ViewPersonBinding) DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.view_person, viewGroup, false));
    }

    @NonNull
    @Override
    protected int getViewModelId(int viewType) {
        return BR.data;
    }

    static class PersonViewHolder extends BaseViewHolder<ViewPersonBinding> {

        public PersonViewHolder(@NonNull ViewPersonBinding binding) {
            super(binding);
        }
    }
}

File: app/src/main/java/com/charlezz/finalarchitecture/feature/local/PersonAdapter.kt
package com.charlezz.finalarchitecture.feature.local

import android.support.v7.util.DiffUtil
import com.charlezz.finalarchitecture.BR
import com.charlezz.finalarchitecture.R
import com.charlezz.finalarchitecture.data.local.Person
import com.charlezz.finalarchitecture.databinding.ViewPersonBinding
import com.charlezz.finalarchitecture.feature.base.BaseAdapter
import com.charlezz.finalarchitecture.feature.base.BaseViewHolder


class PersonAdapter : BaseAdapter<Person, ViewPersonBinding, PersonAdapter.PersonViewHolder>(BR.data,diffCallback) {

    override fun getViewHolderLayoutId(): Int = R.layout.view_person

    companion object {
        val diffCallback = object : DiffUtil.ItemCallback<Person>() {
            override fun areContentsTheSame(oldItem: Person?, newItem: Person?): Boolean {
                return if (oldItem != null && newItem != null){
                    oldItem.id == newItem.id &&
                            oldItem.birth == newItem.birth &&
                            oldItem.name == newItem.name
                } else {
                    false
                }
            }

            override fun areItemsTheSame(oldItem: Person?, newItem: Person?): Boolean {
                return oldItem == newItem
            }
        }
    }
    inner class PersonViewHolder(binding: ViewPersonBinding) : BaseViewHolder<ViewPersonBinding>(binding)
}

